proposals/0106-paritioned-epoch-rewards-pda.md

```diff
@@ -0,0 +1,97 @@
+---
+simd: '0106'
+title: Partitioned Epoch Rewards PDA
+authors:
+  - Tyera Eulberg
+category: Standard
+type: Core
+status: Draft
+created: 2024-01-17
+feature: https://github.com/solana-labs/solana/issues/32166
+extends: 0015
+---
+
+## Summary
+
+Partitioned epoch rewards (SIMD 0015) will split epoch rewards distribution
+across multiple blocks at the start of an epoch. This extension of that SIMD
+proposes storing data about the rewards calculation and partitions in a sysvar
+account at a program-derived address (PDA), which can be queried by clients, or
+indeed within the runtime, to performantly identify the partition for the
+distribution to any particular address.
+
+## Motivation
+
+When we move to partitioned epoch rewards as described in the original SIMD, the
+only way to find the stake or voting rewards for a specific address will be to
+iterate through blocks at the beginning of each epoch, inspecting all the
+rewards. This is because the runtime does not persist information about how the
+rewards were partitioned; in fact, it does not even persist how many blocks the
+rewards distribution spans, so there is no way to predict how long it will take
+or how far through the epoch to go in finding a particular address.
+
+## Alternatives Considered
+
+An alternative to storing partition data in an on-chain account would be to
+record the necessary data in the ledger in some fashion. This could be a
+transaction that gets added to the block, a special RewardType, or metadata that
+gets stored on the node and then duplicated to long-term storage, like Bigtable.
+In fact, it is probably worthwhile to pursue this alternative as well, since it
+will enable finding rewards without access to snapshots or the running chain.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+When partitioned rewards are calculated in the runtime (currently during the
+first block of the epoch), the runtime should populate a PDA that stores the
+partition data needed to recreate the hasher that returns the partition index
+for any address. This data comprises: the number of partitions, the parent
+blockhash, and the hasher kind. More specifically:
+
+```rust
+// Version wrapper to allow future updates
+enum EpochRewardsPartitionDataVersion {
+    V0(PartitionData),
+}
+
+// Extensible list of kinds of hashers used to generate partitions
+enum HasherKind {
+    Sip13,
+}
+
+// Data about rewards partitions for a particular epoch
+struct PartitionData {
+    num_partitions: usize,
+    parent_blockhash: Hash,
+    hasher_kind: HasherKind,
+}
+```
+
+The address of this PDA should include some bytes to prevent griefing (eg.
+`b"EpochRewardsPartitionData"`) and the current epoch number as a little-endian
```

Comment:
ah i think it's actually the _current_ here that's throwing me off... maybe this?
```suggestion
`b"EpochRewardsPartitionData"`) and the rewards distribution epoch number as a little-endian
```

though i don't recall now which we used in the code. i think it'd actually make more sense to seed off the epoch during which the rewards were earned?

Reply:
Yes, your wording suggestion seems better.

> though i don't recall now which we used in the code.

We used the distribution epoch number (`Bank::epoch()`).

> i think it'd actually make more sense to seed off the epoch during which the rewards were earned?

That may be more intuitive from a querying perspective, but I prefer to leave it as-is, since the distributions actually occur within the new epoch. I'll use the distribution epoch with EpochSchedule to find the first slot/block in the epoch.